### PR TITLE
feat: add more integration tests

### PR DESCRIPTION
**Pull request by advasileva**
According to #31

The vast majority of PR is tests, java diff is small)

**Updates:**

+ 6 test cases added
+ EO tests are divided into conversion and compilation (thus, now 17 tests runs separately)
+ 4 tests fails due to one (or more...) bug (they are disabled and todo added)
+ The tests are divided into `fast` (conversion) and `slow` (compilation and run with `eoc`), because the `slow` ones run for 10 or more minutes. The necessary flags added to the `pom`
+ Codecov added

### CODE REVIEW

**Comment by advasileva** [comment]
@yegor256 please take a look

**Comment by yegor256** [comment]
@volodya-lombrozo please, take a look. Maybe it's better to suggest @advasileva to use YAML format for test files, like we do in EO repository? Tests will be more concise, I believe

**Comment by advasileva** [comment]
@yegor256 I was thinking of using YAML for unit tests of transformations, it seemed to me that now the tests are quite long (the largest is 291 lines), so the YAML will be less readable, but I may be wrong

**Comment by yegor256** [comment]
@advasileva the main advantage of YAML is that a test is a _single_ file, which makes it very easy to maintain. Now you have three files per a test -- this is threat to maintainability

**Comment by advasileva** [comment]
@yegor256 hm, makes sense

File: src/test/java/org/eolang/speco/MainTest.java [hunk 1]
```diff
@@ -61,21 +70,52 @@ public final class MainTest {
...
+        );
+    }
+
+    /**
+     * Generates full names of eo test cases.
+     *
+     * @return Collection of test cases names
+     */
+    @SuppressWarnings("PMD.UnusedPrivateMethod")
+    private static Collection<String> getEoTests() {
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/MainTest.java:106]: @advasileva Maybe `cases()` would be better name? What do you think?
https://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]
@advasileva Please, have a look how we write tests for eo programs.

Here is how we do it in dejump project:
1. [PacksTest](https://github.com/objectionary/dejump/blob/master/src/test/java/org/eolang/dejump/PacksTest.java) that allows upload `yaml` files that describe test cases.
2. Here is the folder with [tests](https://github.com/objectionary/dejump/tree/master/src/test/resources/org/eolang/dejump/packs)
3. Here is the concrete [example](https://github.com/objectionary/dejump/blob/master/src/test/resources/org/eolang/dejump/packs/1.1_complex-simple-goto.yaml)

In short, you have 3 sections:
- `xsls` - transformations that you wanna apply
- `before` - initial program before transformation
- `after` - expected program after transformation

I'm using almost the same approach for DeMu project:
- [PacksTest](https://github.com/objectionary/demu/blob/master/src/test/java/org/eolang/demu/PacksTest.java)

But I'm using `jucs` library that simplifies test significantly:
```
<dependency>
  <groupId>org.eolang</groupId>
  <artifactId>jucs</artifactId>
  <version>0.2.0</version>
  <scope>test</scope>
</dependency>
```

File: src/test/java/org/eolang/speco/MainTest.java (hunk 1 above)
Comment (advasileva) [reply src/test/java/org/eolang/speco/MainTest.java]: I agree, but this method is no longer needed and has been removed

**Comment by advasileva** [comment]
@yegor256 @volodya-lombrozo look at the new version of integration tests implemented with .yaml configs. Unit tests with step-by-step checks will be added in another PR

File: .github/workflows/codecov.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+---
+name: codecov
```
Comment (volodya-lombrozo) [inline .github/workflows/codecov.yml:2]: @advasileva Can we move GA into a separate PR, please?

File: src/test/java/org/eolang/speco/SpecoTest.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        }
+    }
+
+    @Disabled
+    @Tag("fast")
+    @ParameterizedTest
+    @ClasspathSource(value = "org/eolang/speco/packs", glob = "**.yaml")
+    public void convertsFromEo(
+        final String pack,
+        @TempDir(cleanup = CleanupMode.NEVER) final Path temp) throws IOException {
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:89]: @advasileva Do you need `cleanup = CleanupMode.NEVER`? It seems that it will leave garbage after each test run

File: src/test/java/org/eolang/speco/SpecoTest.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                ),
+                Files.readAllLines(temp.resolve(path.getFileName())),
+                Matchers.equalTo(
+                    Files.readAllLines(path)
+                )
+            );
+        }
+    }
+
+    @Disabled
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:83]: @advasileva I would be better to add a puzzle that describes why "@Disable" annotation is here and when we have to remove it

File: src/test/java/org/eolang/speco/SpecoTest.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        }
+        final Process process = new ProcessBuilder(
+            executor,
+            flag,
+            String.format("eoc link -s %s && eoc --alone dataize app && eoc clean", target)
+        ).start();
+        final StringWriter writer = new StringWriter();
+        IOUtils.copy(process.getInputStream(), writer);
+        final String[] output = writer.toString().split("\\r?\\n");
+        return Arrays.copyOfRange(output, 11, output.length - 1);
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:163]: @advasileva `11` is a magic number. It's better to create a static final field and add `JavaDoc` for it with explanation "why it is exactly 11".

File: src/test/java/org/eolang/speco/SpecoTest.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    @ClasspathSource(value = "org/eolang/speco/packs", glob = "**.yaml")
+    public void compilesFromEo(
+        final String pack,
+        @TempDir(cleanup = CleanupMode.NEVER) final Path temp) throws IOException {
+        final Map<String, Object> script = new Yaml().load(pack);
+        MatcherAssert.assertThat(
+            "Unexpected execution result",
+            SpecoTest.exec(SpecoTest.runSpeco(script, temp).toString()),
+            Matchers.equalTo(
+                script.get("result").toString().split("\\r?\\n")
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:111]: @advasileva Maybe we can use `System.lineSeparator()` here instead?

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]
@advasileva We have to leave the system state the same after all tests (they don't have to leave any side effects): please, try to avoid using `@TempDir(cleanup = CleanupMode.NEVER)`. 

And even that PR could be divided into many small separate PR`s (according with your bullet list in the description).

File: .github/workflows/codecov.yml (hunk 2 above)
Comment (advasileva) [reply .github/workflows/codecov.yml:2]: @volodya-lombrozo Done in #41

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 3 above)
Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java:89]: @volodya-lombrozo This policy [fixes](https://github.com/objectionary/speco/pull/32/commits/95d2f9a29f84df2fd4ea35117b9b32be3c711a14) tests failures on windows with Java 11, [e.g.](https://github.com/objectionary/speco/actions/runs/4097463770/jobs/7065856235)

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 6 above)
Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java]: @volodya-lombrozo Yes, done

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 3 above)
Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java:102]: I tried several ways to fix it, but it didn't work out. The problem is only on Windows for Java 11.

File: .github/workflows/codecov.yml (hunk 2 above)
Comment (volodya-lombrozo) [reply .github/workflows/codecov.yml:2]: @advasileva What do you think if we remove that code from the current PR since it's already in other PR?

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 6 above)
Comment (volodya-lombrozo) [reply src/test/java/org/eolang/speco/SpecoTest.java:124]: @advasileva Thanks

File: src/test/java/org/eolang/speco/SpecoTest.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    /**
+    * Compiles EO program.
+    *
+    * @param target Path to the dir with target EO program
+    * @return List of lines in output
+    * @throws IOException Iff IO error
+    */
+    private static String[] exec(final String target) throws IOException {
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:158]: @advasileva Maybe the better name would be `compile` or `dataize`, what do you think?

File: src/test/java/org/eolang/speco/SpecoTest.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+    /**
+     * Runs Speco.
+     *
+     * @param script Yaml data object
+     * @param temp Path to the temporary dir
+     * @return Path to the output dir
+     * @throws IOException Iff IO error
+     */
+    private static Path runSpeco(final Map<String, Object> script, final Path temp)
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:137]: @advasileva Here we can just leave `run`

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]

File: src/test/java/org/eolang/speco/SpecoTest.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+            executor = "bash";
+            flag = "-c";
+        }
+        final Process process = new ProcessBuilder(
+            executor,
+            flag,
+            String.format("eoc link -s %s && eoc --alone dataize app && eoc clean", target)
+        ).start();
+        final StringWriter writer = new StringWriter();
+        IOUtils.copy(process.getInputStream(), writer);
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:174]: @advasileva Since you are using `InputStream` here, I highly recommend to close it and check if you leave any other open files after tests, it could be a cause of problem with JUnit temporary folders.

Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:174]: @advasileva Also `IOUtils` is deprecated as far as I know.

File: src/test/java/org/eolang/speco/SpecoTest.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        final String[] output = writer.toString().split("\\r?\\n");
```
Comment (volodya-lombrozo) [inline src/test/java/org/eolang/speco/SpecoTest.java:175]: @advasileva Could we use `System.lineSeparator()` here?

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 10 above)
Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java:175]: @volodya-lombrozo I tried use it, but commit with it [failed on windows](https://github.com/objectionary/speco/actions/runs/4107654594/jobs/7087404723), so I [return regex](https://github.com/objectionary/speco/pull/32/commits/77f32e30c5555841fbb72208f67a6e985a58bbc6)

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 7 above)
Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java:158]: @volodya-lombrozo I named the method `exec` (maybe `execute` is better) because it returns the output of the program along the specified path. It seems to me that compilation and dataization are intermediate steps that do not reflect the purpose of the method. If you think I'm wrong, then I'll rename it

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 9 above)
Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java:168]: Yes, `Charset` added (version with it is not deprecated)

Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java]: @volodya-lombrozo Closing the streams in SpecoTest did not help, I will try to close the files

Comment (volodya-lombrozo) [reply src/test/java/org/eolang/speco/SpecoTest.java]: @advasileva I also highly recommend to use [Process.waitFor](https://docs.oracle.com/javase/8/docs/api/java/lang/Process.html#waitFor--) method.

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 7 above)
Comment (volodya-lombrozo) [reply src/test/java/org/eolang/speco/SpecoTest.java:153]: @advasileva the names `exec` and even `execute` don't say anything about their purpose (especially about "returns the output of the program along the specified path"). The main purpose of that method is dataization as I see it from the code. I believe it would be better to rename.

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]

File: src/test/java/org/eolang/speco/SpecoTest.java (hunk 9 above)
Comment (volodya-lombrozo) [reply src/test/java/org/eolang/speco/SpecoTest.java]: @advasileva As I see from logs we have that exception most of the time:
```
SpecoTest.compilesFromEo(String, Path)[3] � IO Failed to delete temp directory C:\Users\RUNNER~1\AppData\Local\Temp\junit18001172281059271577. The following paths could not be deleted (see suppressed exceptions for details): , input_prs_aoi
```

This files are created by that method:
```java
 private static Path parse(final Path input) throws IOException {
        final StringBuilder name = new StringBuilder(input.toString());
        final Path source = Path.of(name.append("_prs").toString());
        FileUtils.copyDirectory(input.toFile(), source.toFile());
        for (final Path path : Files.newDirectoryStream(source)) {
            new Syntax(
                "scenario",
                new InputOf(String.format("%s\n", Files.readString(path))),
                new OutputTo(new FileOutputStream(path.toFile()))
            ).parse();
        }
        LauncherKt.launch(source.toString());
        return Path.of(name.append("_aoi").toString());
    }
```

Maybe we have also close  `Files.newDirectoryStream(source)` stream. It's just a guess.

Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java]: I [debug](https://github.com/advasileva/speco/pull/6) this problem, it turned out that `LauncherKt.launch(source.toString());` is a process that does not allow deleting files. I do not know how to fix this from java (`FORCE` removal did not help), so I suggest refuse to use `@TempDir` or set a `cleanup = CleanupMode.NEVER` policy

Comment (advasileva) [reply src/test/java/org/eolang/speco/SpecoTest.java]: @volodya-lombrozo check out the message above

**Comment by advasileva** [comment]
@volodya-lombrozo I discussed this bug with Yegor, this problem is [issue of junit5](https://github.com/junit-team/junit5/issues/2811). Now I suppressed it with `@DisableOnOs(OS.Windows)` annotation, what do you think?

**Comment by volodya-lombrozo** [comment]
> @volodya-lombrozo I discussed this bug with Yegor, this problem is [issue of junit5](https://github.com/junit-team/junit5/issues/2811). Now I suppressed it with `@DisableOnOs(OS.Windows)` annotation, what do you think?

Looks good, but I would add a puzzle to investigate the problem with `@DisableOnOs(OS.Windows)`

**Comment by advasileva** [comment]
> > @volodya-lombrozo I discussed this bug with Yegor, this problem is [issue of junit5](https://github.com/junit-team/junit5/issues/2811). Now I suppressed it with `@DisableOnOs(OS.Windows)` annotation, what do you think?
> 
> @advasileva, it looks good, but I would add a puzzle to investigate the problem with `@DisableOnOs(OS.Windows)`

@volodya-lombrozo, done

**Comment by volodya-lombrozo** [comment]
@advasileva Thanks.
@yegor256 Could you merge that changes, please?

**Comment by advasileva** [comment]
@volodya-lombrozo, thanks for review

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0pdd** [comment]
@advasileva the puzzle [#54](https://github.com/objectionary/speco/issues/54) is still not solved.
